tools/dig/dig.1
.\"
.\" Distributed with 'dig' version 1.0 from University of Southern
.\" California Information Sciences Institute (USC-ISI).
.\"
.\"       dig.1   1.0 (USC-ISI) 4/15/89
.\"
.TH DIG 1 "April 15, 1989"
.SH NAME
dig \- send domain name query packets to name servers
.SH SYNOPSIS
.IP "dig domain"
.ns
.IP "    [@<server>]"
.ns
.IP "    [<domain query type>]"
.ns
.IP "    [<domain query class>]"
.ns
.IP "    [+<query parameter>]"
.ns
.IP "    [-<dig parameter>]"
.ns
.IP "    [%comment-ignored]"
.sp 1
.SH DESCRIPTION
The Domain Name System, currently used by a majority of the
Internet's component networks, is a large and complex
autonomously administrated distributed database which provides
a network wide name service. It's size and the autonomously 
administered nature of the system make it an ideal breeding
ground for misconfiguration and other problems.
.sp 1
Dig (domain information groper) is a flexible command line tool 
which can be used to gather information from the Domain
Name System servers. Dig has two modes: simple interactive mode
which makes a single query, and batch which produces a query for
a list of several queries. All query options are accessible from
the command line.
.sp 1
The usual simple use of 'dig' will take the form:
.sp 1
	dig @server_name domain_name dns_type dns_class
.sp 1
where:
.sp 1
Server_name - may be either a domain name or a dot-notation
Internet address. If this optional field is omitted, 'dig'
will attempt to use the default name server for your machine.
.sp 1
 NOTE: If a domain name is specified, this will be resolved
 using the domain name system resolver (i.e. bind). If your
 system does not support DNS,  you may *have* to specify a
 dot-notation address.  Alternatively, if there is a server
 at your disposal somewhere,  all that is required is that
 /etc/resolv.conf be present and indicate where the default
 name servers  reside,  so that  server_name itself can be
 resolved. See man resolver(5) for /etc/resolv.conf infor-
 mation.  (WARNING: Changing /etc/resolv.conf will affect
 the standard  resolver library and  potentially several
 programs which use it.) As an option, the user may set the
 environment variable LOCALRES to name a file which is to
 be used instead of /etc/resolv.conf (LOCALRES is specific
 to the 'dig' resolver and  not referenced by the standard
 resolver). If the LOCALRES variable is not set or the file
 is not readable then /etc/resolv.conf will be used.
.sp 2
Domain_name - is the domain for which you are requesting
information.
.sp 2
Dns_type - is the type of information (DNS query type) that
you are requesting. If omitted, the default is "a" (T_A = address).
.IP "	a       T_A     network address"
.ns
.IP "	any     T_ANY   all/any information about specified domain"
.ns
.IP "	mx      T_MX    mail exchanger for the domain"
.ns
.IP "	ns      T_NS    name servers"
.ns
.IP "	soa     T_SOA   zone of authority record"
.ns
.IP "	hinfo   T_HINFO host information"
.ns
.IP "	txt     T_TXT   arbitrary number of strings"
.ns
.IP "                     (not yet supported by bind)"
.ns
.IP "	..."
.ns
.IP "             (see RFC 1035 for complete list)"
.sp 1
.PP
Dns_class - is the network class requested in the query. If
omitted, the default is "in" (C_IN = Internet).
.IP "	in      C_IN    Internet class domain"
.ns
.IP "	any     C_ANY   all/any class information"
.ns
.IP "	..."
.ns
.IP "	        (see RFC 1035 for complete list)"
.sp 1
NOTE:
  "Any" can be used to specify a class and/or a type of
   query. Dig will parse the first occurrence of "any"
   to mean query_type = T_ANY. To specify query_class =
   C_ANY you must either specify "any" twice, or set
   query_class using "-c" option (see below).
.sp 1
.SH OTHER OPTIONS
%ignored-comment
   "%" is used to included an argument that is simply not
   parsed.  This may be useful  if running 'dig' in batch
   mode. Instead of resolving every @server-domain-name in
   a list of queries, you can avoid the overhead of doing
   so, and still have the domain name on the command line
   as a reference. Example:
.sp 1
	dig @128.9.0.32 %venera.isi.edu mx isi.edu
.sp 2
-<dig option>
   "-" is used to specify an option which effects the
   operation of dig. The following options are currently
   available (although not guaranteed to be useful):
.sp 1
   -f XXX  File for dig batch mode. File XXX contains a list
           of query specifications (dig command lines) which
           are to be executed successively. Lines beginning
           with ';', '#', or '\\n' are ignored. Other options
           may still appear on command line, and will be in
           effect for each batch query.
.sp 1
   -T XXX  Time in seconds between start of successive
           queries when running in batch mode. Can be used
           to keep two or more batch dig commands running
           roughly in sync. Default is zero.
.sp 1
   -p XXX  Port number. Query a name server listening to a
           non-standard port number.
.sp 1
   -P["ping string"]
           After query returns, execute a ping(8) command
           for response time comparison. This rather
           unelegantly makes a call to the shell. The last
           three lines of statistics is printed for the
           command:
.sp 1
			ping -s server_name 56 3
.sp 1
           If the optional "ping string" is present, it
           replaces "ping -s" in the shell command. A
           server must be specified on the command line.
.sp 1
   -t XXX  Specify type of query. May specify either an
           integer value to be included in the type field
           or use the abbreviated mnemonic as discussed
           above (i.e. mx  = T_MX).
.sp 1
   -c XXX  Specify class of query. May specify either an
           integer value to be included in the class field
           or use the abbreviated mnemonic as discussed
           above (i.e. in = C_IN).
.sp 1
   -envsav This flag specifies that the environment after
           all of the arguments are parsed should be saved
           to a file to become the default environment.
           Useful if you do not like the standard set of
           defaults and do not desire to include such a
           large number of options each time 'dig' is used.
           The environment consists of resolver state
           variable flags, timeout, retries as well as
           the flags detailing 'dig' output (see below).
           If the environment variable LOCALDEF is set
           to the name of a file, this is where the default
           environment is saved. If not, the file "DiG.env"
           is created in the current working directory.
.sp 1
           NOTE: LOCALDEF is specific to the 'dig' resolver,
           and will not affect operation of the standard
           resolver library.
.sp 1
           Each time 'dig' is executed, a check is made
           for LOCALDEF. If it is defined and the file
           is readable, then the environment is restored
           before any arguments are parsed.
.sp 1
   -[no]stick
           This flag is only useful in batch queries.
           It indicates  whether (or not)  to restore the
           default environment (see -envsav) before parsing
           and sending each query. The default environment
           is  either the  standard defaults, those  from
           LOCALDEF,  or those  set by  "-envset".
           The default is "-nostick".
.sp 1
   -envset This flag in conjunction with "-stick" may be
           useful for batch query runs. When "-envset" is
           specified, the environment after the arguments
           are parsed becomes the default environment for
           the duration of the batch file.
.sp 2
+<query option>
   "+" is used to specify an option to be changed in the
   query packet or to change 'dig' output specifics. Many
   of these are the same parameters accepted by 'nslookup'.
   If an option requires a parameter, the form is as
   follows:
.sp 1
	         +keyword[=value]
.sp 1
   Most keywords can be abbreviated.  Parsing of the "+"
   options  is very  simplistic  --  A value must not be
   separated from its keyword by white space. The following
   keywords are currently available:
.sp 1
.IP "      Valid keywords:      Meaning:"
.IP "      (abbreviation)      [default]"
.IP "      ---------------     ---------"
.sp 1
.IP "      [no]debug (deb)    turn on/off debugging mode [deb]"
.ns
.IP "      [no]d2             turn on/off extra debugging mode [nod2]"
.ns
.IP "      [no]recurse (rec)  use/don't use recursive lookup [rec]"
.ns
.IP "      retry=#  (ret)     set number of retries to # [4]"
.ns
.IP "      time=#   (ti)      set timeout length to # seconds [4]"
.ns
.IP "      [no]ko             keep open option (implies vc) [noko]"
.ns
.IP "      [no]vc             use/don't use virtual circuit [novc]"
.ns
.IP "      [no]defname (def)  use/don't use default domain name [def]"
.ns
.IP "      [no]search  (sea)  use/don't use domain search list [sea]"
.ns
.IP "      domain=NAME (do)   set default domain name to NAME"
.ns
.IP "      [no]ignore  (i)    ignore/don't ignore trunc. errors [noi]"
.ns
.IP "      [no]primary (pr)   use/don't use primary server [nopr]"
.ns
.IP "      [no]aaonly  (aa)   authoritative query only flag [noaa]"
.ns
.IP "      [no]sort    (sor)  sort resource records [nosor]"
.ns
.IP "      [no]cmd            echo parsed arguments [cmd]"
.ns
.IP "      [no]stats   (st)   print query statistics (RTT,etc) [st]"
.ns
.IP "      [no]qr             print outgoing query [noqr]"
.ns
.IP "      [no]reply   (rep)  print reply [rep]"
.ns
.IP "      [no]header  (he)   print certain parts of header [he]"
.ns
.IP "      [no]Header  (H)    print all/any of header [H]"
.ns
.IP "      [no]ttlid   (tt)   print TTLs and pkt ID's [tt]"
.ns
.IP "      [no]ques    (qu)   print question section [qu]"
.ns
.IP "      [no]answer  (an)   print answer section [an]"
.ns
.IP "      [no]author  (au)   print authoritative section [au]"
.ns
.IP "      [no]addit   (ad)   print additional section [ad]"
.ns
.IP "      pfdef              set to default print flags"
.ns
.IP "      pfmin              set to minimal default print flags"
.ns
.IP "      pfset=#            set print flags to #"
.ns
.IP "      pfand=#            bitwise and print flags with #"
.ns
.IP "      pfor=#             bitwise or print flags with #"
.sp 2
retry/time -- the retransmission strategy used by resolver library
when sending datagram queries is as follows. (Note: the number of 
servers and/or addresses is always one when using 'dig').
.IP "	   for i = 0 to retry-1"
.ns
.IP "	       for j = 1 to num_servers"
.ns
.IP "	         send_query"
.ns
.IP "	         wait((time * (2**i)) / num_servers)"
.ns
.IP "	        rof"
.ns
.IP "	   rof"
.sp 1
pfset/pfand/pfor - these were included to make manipulation of
the various print options less tedious. The currently defined
values are:
.IP "	    PRF_STATS    0x0001    RTT, query host, server host info"
.ns
.IP "	    PRF_CMD      0x0008    dig command line echo"
.ns
.IP "	    PRF_QUES     0x0010    questions section"
.ns
.IP "	    PRF_ANS      0x0020    answers section"
.ns
.IP "	    PRF_AUTH     0x0040    authoritative section"
.ns
.IP "	    PRF_ADD      0x0080    additional records section"
.ns
.IP "	    PRF_HEAD2    0x0200    header flags, section RR counts"
.ns
.IP "	    PRF_TTLID    0x0400    ttl and packet id number"
.ns
.IP "	    PRF_HEADX    0x0700    any/all packet header info"
.ns
.IP "	    PRF_QUERY    0x1000    outgoing query packet info"
.ns
.IP "	    PRF_REPLY    0x2000    reply packet info"
.ns
.IP "	    PRF_SORT     0x8000    sort various response sections"
.ns
.IP "	    PRF_DEF      0x27f9    default dig settings"
.ns
.IP "	    PRF_MIN      0xa133    minimalistic dig settings for"
.ns
.IP "	                           (future) automated server testing"
.sp 1
When setting the print options, if you want to see information other
than statistics, you should choose to examine the outgoing (0x1000), 
incoming (0x2000), or both packets plus the specific sections of the packet
you are interested in.
.sp 1
.SH DETAILS
Dig requires a slightly modified version of the bind resolver(3)
library to gather count and time statistics. Otherwise, it is
straight-forward (albeit not pretty) effort of parsing arguments
and setting appropriate parameters. Dig uses resolver routines
res_init(), res_mkquery(), res_send() as well as accessing _res
structure. Compiling 'dig' with the standard resolver library
is possible, but will change the output format, make the print
options meaningless, and not gather RTT and packet count stats.
.sp 1
.SH FILES
.IP "o /etc/resolv.conf \- initial domain name and"
.ns
.IP "  name server addresses"
.ns
.IP "o environment var LOCALRES names file to use in place of
.ns
.IP "  /etc/resolv.conf"
.ns
.IP "o environment var LOCALDEF names default environment file"
.SH AUTHOR
Steve Hotz 
hotz@isi.edu
.sp 1
.SH ACKNOWLEDGEMENTS
Dig uses functions from nslookup(1) authored by Andrew Cherenson;
taken from Berkeley bind 4.8 distribution. The resolver library
is primarily from the Berkeley 4.8 bind distribution.
.sp 1
.SH BUGS
When running 'dig' in batch mode, if an error occurs in one of the
resolver routines 'dig' exits. The preferred behavior would be to
give up only for that particular query and continue with the
query list. The fix involves modifying the resolver routines to
return a status indicating error, rather than simply exiting.
.sp 1
The ping option simply makes a system() call to the shell.
This should be replaced with internal ping code.
.sp 1
.SH SEE ALSO
.sp 1
named(8)  resolver(3)  resolver(5)  nslookup(1)


CHANGES
This is version 4.8.2 of bind.  It is a test release that includes
versions 4.8 and 4.8.1 with fixes.  Although it is currently described
as a test release, it is believed to be reasonably stable and more
usable than the currently-released versions 4.8 and 4.8.1.  Here are
some of the more important changes:

  o Support has been added for the the T_TXT data type and for the class
    C_HS.  These are both used by Hesiod from Project Athena at MIT.

  o All of the pathnames have been put into one header file.  This
    makes it easier to change the location based upon your local
    configuration.

  o Numerous bugs have been fixed:  Adding a new authoritative zone now
    works when the server has a cached SOA record.  Comparisons in the
    db now look at type and class as well instead of dropping records
    with identical data.  Scheduling of maintenance interrupts has been
    moved to one routine avoid spurious ones.  Named goes into the background
    after more of the initialization is done.  Stream connection queue
    handling was cleaned up including a bug that caused data corruption
    and core dumps.  Sys5 no longer can have multiple transfers of the
    same zone occuring at the same time.  Handle CNAME -> CNAME loops
    more gracefully.  Avoid making one server never get queried.

There are several bug reports that have yet to be integrated into this
version.  Hopefully they will be dealt with in the next release.  Please
send feedback on this release.

Notes from versions 4.8.1 and 4.8 follow:

------------------

This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.


tools/dig/dig.c


 /*******************************************************************
 **      DiG -- Domain Information Groper                          **
 **                                                                **
 **        dig.c - Version 1.0                                     **
 **        Developed by: Steve Hotz & Paul Mockapetris             **
 **        USC Information Sciences Institute (USC-ISI)            **
 **        Marina del Rey, California                              **
 **        1989                                                    **
 **                                                                **
 **     DiG is Public Domain, and may be used for any purpose as   **
 **     long as this notice is not removed.                        **
 *******************************************************************/

#include "hfiles.h"

#include <sys/types.h>
#include <netinet/in.h>
#include <stdio.h>
#include RESOLVH
#include <sys/file.h>
#include <sys/stat.h>
#include <ctype.h> 
#include "subr.c"
#include "strcasecmp.c"
#include "pflag.h"
#include "options.c"

#ifdef RESLOCAL
#include "qtime.c"
#else
#include "qtime.h"
#endif

int pfcode = PRF_DEF;
extern char *inet_ntoa();
extern struct state _res;

FILE  *qfp;

#define MAXDATA		256   
#define SAVEENV "DiG.env"

char defsrv[]="default";


 /*
 ** Take arguments appearing in simple string (from file)
 ** place in char**.
 */
stackarg(y,l)
     char *l;
     char **y;
{
  int done=0;
  while (!done) {
    switch (*l) {
    case '\t':
    case ' ':  l++;    break;
    case NULL:
    case '\n': done++; *y = NULL; break;
    default:   *y++=l;  while (!isspace(*l)) l++;
      if (*l == '\n') done++; *l++ = NULL; *y = NULL;
}}}


main(argc, argv)
     int argc;
     char **argv;
{
	struct hostent *hp;
	short port = htons(NAMESERVER_PORT);
	char packet[PACKETSZ];
	char answer[PACKETSZ];
	int n;
	char doping[90];
	char pingstr[50];
	char *afile;
	unsigned long tmpaddr;

	struct timeval exectime, tv1,tv2,tv3;
	char curhost[30];

	char *srv;
	int anyflag = 0;
	int sticky = 0;
	int tmp; 
	int qtype = 1, qclass = 1;
	int addrflag = 0;

	char cmd[256];
	char domain[128];
	char **vtmp;
	char *args[30];
	char **ax;
	char **ay;
	int once = 1, dofile=0; /* batch -vs- interactive control */
	char fileq[100];
	char *qptr;
	int  fp;
	int wait=0;
	int envset=0, envsave=0;
	int Xpfcode, Tpfcode, Toptions;
	char *Xres, *Tres;


	res_init();
	_res.options |= RES_DEFAULT;
	Xres = (char *) malloc(sizeof(_res)+1);
	Tres = (char *) malloc(sizeof(_res)+1);

 /*
 ** If LOCALDEF in environment, should point to file
 ** containing local favourite defaults.
 */
	if (((afile= (char *) getenv("LOCALDEF")) != (char *) NULL) &&
	    ((fp=open(afile,O_RDONLY)) > 0)) {
	  read(fp,Xres,sizeof(struct state));
	  read(fp,&Xpfcode,sizeof(int));
	  close(fp);
	  bcopy(Xres,&_res,sizeof(struct state));
	  pfcode = Xpfcode;
	}
 /*
 **   check for batch-mode DiG 
 */
	vtmp = argv; ax=args;
	while (*vtmp != NULL) {
	  if (strcmp(*vtmp,"-f",2) == 0) {
	    dofile++; once=0;
	    if ((qfp = fopen(*++vtmp,"r")) == NULL) {
	      fflush(stdout);
	      perror("file open");
	      fflush(stderr);
	      exit(1);
	    }
	  } else {
	    *ax++ = *vtmp;
	  }
	  vtmp++;
	}

	_res.id = 1;
	gettimeofday(&tv1,NULL);

 /*
 **  Main section: once if cmd-line query
 **                while !EOF if batch mode
 */
	*fileq=NULL;
	while ((dofile && (fgets(fileq,100,qfp) !=NULL)) || 
	       ((!dofile) && (once--))) 
	  {
	    if ((*fileq=='\n') || (*fileq=='#') || (*fileq==';'))
	      continue; /* ignore blank lines & comments */

/*
 * "sticky" requests that before current parsing args
 * return to current "working" environment (X******)
 */
	    if (sticky) {
	      bcopy(Xres,&_res,sizeof(struct state));
	      pfcode = Xpfcode;
	    }

/* concat cmd-line and file args */
	    ay=ax;
	    qptr=fileq;
	    stackarg(ay,qptr);

	    /* defaults */
	    qtype=qclass=1;
	    *pingstr=0;
	    srv=NULL;

	    sprintf(cmd,"\n; <<>> DiG <<>> ");
	    argv = args;
/*
 * More cmd-line options than anyone should ever have to
 * deal with ....
 */
	    while (*(++argv) != NULL) { 
	      strcat(cmd,*argv); strcat(cmd," ");
	      if (**argv == '@') {
		srv = (*argv+1);
		continue;
	      }
	      if (**argv == '%')
		continue;
	      if (**argv == '+') {
		SetOption(*argv+1);
		continue;
	      }
	 
	      if (strcmp(*argv,"-nost",5) == 0) {
		sticky=0;; continue;
	      } else if (strcmp(*argv,"-st",2) == 0) {
		sticky++; continue;
	      } else if (strcmp(*argv,"-envsa",5) == 0) {
		envsave++; continue;
	      } else if (strcmp(*argv,"-envse",5) == 0) {
		envset++; continue;
	      }

         if (**argv == '-') {
	   switch (argv[0][1]) { 
	   case 'T': wait = atoi(*++vtmp); break;
	   case 'c': 
	     if ((tmp = atoi(*++argv)) || *argv[0]=='0') {
	       qclass = tmp;
	     } else if (tmp = StringToClass(*argv,0)) {
	       qclass = tmp;
	     } else {
	       printf("; invalid class specified\n");
	     }
	     break;
	   case 't': 
	     if ((tmp = atoi(*++argv)) || *argv[0]=='0') {
	       qtype = tmp;
	     } else if (tmp = StringToClass(*argv,0)) {
	       qtype = tmp;
	     } else {
	       printf("; invalid type specified\n");
	     }
	     break;
	   case 'p': port = htons(atoi(*++argv)); break;
	   case 'P':
	     if (argv[0][2] != NULL)
	       strcpy(pingstr,&argv[0][2]);
	     else
	       strcpy(pingstr,"ping -s");
	     break;
	   } /* switch - */
	   continue;
	 } /* if '-'   */

	      if ((tmp = StringToType(*argv,-1)) != -1) { 
		if ((T_ANY == tmp) && anyflag++) { 	
		  qclass = C_ANY; 	
		  continue; 
		} 
		qtype = tmp; 
	      }
	      else if ((tmp = StringToClass(*argv,-1)) != -1) { 
		qclass = tmp; 
	      }	 else {
		bzero(domain,128);
		sprintf(domain,"%s",*argv);
	      }
	    } /* while argv remains */
if (pfcode & 0x80000)
  printf("; pflag: %x res: %x\n", pfcode, _res.options);
	  
/*
 * Current env. (after this parse) is to become the
 * new "working environmnet. Used in conj. with sticky.
 */
	    if (envset) {
	      bcopy(&_res,Xres,sizeof(struct state));
	      Xpfcode=pfcode;
	      envset=0;
	    }

/*
 * Current env. (after this parse) is to become the
 * new default saved environmnet. Save in user specified
 * file if exists else is SAVEENV (== "DiG.env").
 */
	    if (envsave) {
	      if ((((afile= (char *) getenv("LOCALDEF")) != (char *) NULL) &&
		   ((fp=open(afile,O_WRONLY|O_CREAT|O_TRUNC,
			     S_IREAD|S_IWRITE)) > 0)) ||
		  ((fp = open(afile,SAVEENV,O_WRONLY|O_CREAT|O_TRUNC,
			       S_IREAD|S_IWRITE)) > 0)) {
		write(fp,&_res,sizeof(struct state));
		write(fp,&pfcode,sizeof(int));
		close(fp);
	      }
	      envsave=0;
	    }

	    if (pfcode & PRF_CMD)
	      printf("%s\n",cmd);

      addrflag=anyflag=0;

/*
 * Find address of server to query. If not dot-notation, then
 * try to resolve domain-name (if so, save and turn off print 
 * options, this domain-query is not the one we want. Restore
 * user options when done.
 * Things get a bit wierd since we need to use resolver to be
 * able to "put the resolver to work".
 */

   if (srv != NULL) {
     tmpaddr = inet_addr(srv);
     if (tmpaddr != (unsigned)-1) {
       _res.nscount = 1;      
       _res.nsaddr.sin_addr.s_addr = tmpaddr;
     } else {
      Tpfcode=pfcode;
      pfcode=0;
      bcopy(&_res,Tres,sizeof(_res));
      Toptions = _res.options;
      _res.options = RES_DEFAULT;
      res_init();
      hp = gethostbyname(srv);
      pfcode=Tpfcode;
      if (hp == NULL) {
	fflush(stdout);
	perror("Bad server: using default nameserver and timer opts.\n");
	fflush(stderr);
	_res.options = Toptions;
      } else {
	bcopy(Tres,&_res,sizeof(_res));
	bcopy(hp->h_addr, &_res.nsaddr_list[0].sin_addr, hp->h_length);
	_res.nscount = 1;
      }
    }
     _res.id += _res.retry;
   }


       _res.id += _res.retry;
       _res.options |= RES_DEBUG;
       _res.nsaddr.sin_family = AF_INET;
       _res.nsaddr.sin_port = port;

       n = res_mkquery(QUERY, domain, qclass, qtype, (char *)0, 0,
		       NULL, packet, sizeof(packet));
       if (n < 0) {
	 fflush(stderr);
	 printf(";; res_mkquery: buffer too small\n\n");
	 exit(-1);
       }

       if ((n = res_send(packet, n, answer, sizeof(answer))) < 0) {
	 fflush(stdout);
	 perror(";; res_send");
	 fflush(stderr);

#ifndef RESLOCAL
/*
 * resolver does not currently calculate elapsed time
 * if there is an error in res_send()
 */
/*
	 if (pfcode & PRF_STATS) {
	   printf(";; Error time: "); prnttime(&hqtime); putchar('\n');
	 }
*/
#endif
       }
#ifndef RESLOCAL
       else {
	 if (pfcode & PRF_STATS) {
	   printf(";; Sent %d pkts, answer found in time: ",hqcount);
	   prnttime(&hqtime);
	   if (hxcount)
	     printf(" sent %d too many pkts",hxcount);
	   putchar('\n');
	 }
       }
#endif

       if (pfcode & PRF_STATS) {
	 gettimeofday(&exectime,NULL);
	 gethostname(curhost,30);
	 printf(";; FROM: %s to SERVER: %s\n",curhost,(srv==NULL)?defsrv:srv);
	 printf(";; WHEN: %s",ctime(&(exectime.tv_sec)));
       }
	  
       fflush(stdout);
/*
 *   Argh ... not particularly elegant.
 *   Should put in *real* ping code.
 */
       if (*pingstr) {
	 sprintf(doping,"%s %s 56 3 | tail -3",pingstr,(srv==NULL)?defsrv:srv);
	 system(doping);
       }
       putchar('\n');

/*
 * Fairly crude method and low overhead method of keeping two
 * batches started at different sites somewhat synchronized.
 */
       gettimeofday(&tv2);
       tv1.tv_sec += wait;
       difftime(&tv1,&tv2,&tv3);
       if (tv3.tv_sec > 0)
	 sleep(tv3.tv_sec);
  }
}

